### PR TITLE
Support constant byte array literal

### PR DESCRIPTION
**Pull request by ushirask**
Fixes https://github.com/ballerina-platform/ballerina-lang/issues/40536
Fixes https://github.com/ballerina-platform/ballerina-lang/issues/37423

## Check List 
- [ ] Read the [Contributing Guide](https://github.com/ballerina-platform/ballerina-lang/blob/master/CONTRIBUTING.md)
- [ ] Updated Change Log
- [ ] Checked Tooling Support (#<Issue Number>)
- [ ] Added necessary tests
  - [ ] Unit Tests
  - [ ] Spec Conformance Tests
  - [ ] Integration Tests
  - [ ] Ballerina By Example Tests
- [ ] Increased Test Coverage   
- [ ] Added necessary documentation  
  - [ ] API documentation 
  - [ ] Module documentation in Module.md files
  - [ ] Ballerina By Examples

### CODE REVIEW

File: tests/jballerina-unit-test/src/test/resources/test-src/types/constant/constant-assignment-negative.bal [hunk 1]
```diff
@@ -24,3 +24,5 @@ const Foo F1 = {x: {b : "a"}, i: 1, c: 2};
 const record{|X x; int i;|} F2 = {x: {b : "a"}, i: 1, c: 2};
 
 const X F3 = {b : "b"};
+
+const string[] Y = base16 `aabb`;
```
Comment (prakanth97) [inline tests/jballerina-unit-test/src/test/resources/test-src/types/constant/constant-assignment-negative.bal:28]: Shall we add a negative test for assigning to a tuple type?

File: compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ConstantTypeChecker.java [hunk 2]
```diff
@@ -255,6 +255,21 @@ public void visit(BLangLiteral literalExpr, AnalyzerData data) {
...
+            BType expType = Types.getReferredType(data.expType);
+            if (expType.tag == TypeTags.ARRAY && ((BArrayType) expType).state == BArrayState.INFERRED) {
+                ((BArrayType) expType).size = listConstructorExpr.exprs.size();
+                ((BArrayType) expType).state = BArrayState.CLOSED;
+            }
+            if (types.typeIncompatible(literalExpr.pos, listConstructorExpr.getBType(), data.expType)) {
+                data.resultType = symTable.semanticError;
+                return;
+            }
+            data.resultType = checkListConstructorCompatibility(literalType, listConstructorExpr, data);
```
Comment (prakanth97) [inline compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ConstantTypeChecker.java:269]: As far as I understood, first we create a `BLangListConstructorExpr ` and use that return tuple type here. But can't we directly create tuple type at #L259. It seems unnecessary step to me. Is there any use of the created ListConstructorExpr other than passing it to this method?

Comment (ushirask) [reply compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/ConstantTypeChecker.java:269]: +1, we should be able return the tuple type and just check type compatibility. Will update.

**Comment by prakanth97** [comment]
> Since we know the constant value at compile-time, we should support below case
> 
> ```
> const [170] X = base16 `aa`;
> ```

[To fix this ](https://github.com/prakanth97/ballerina-lang/commit/f46075cfed4a07b12989f49608c9dfb00679c69d) With this changes this issue could be fixed. Could you pls add those changes?

File: tests/jballerina-unit-test/src/test/resources/test-src/types/constant/constant-assignment-negative.bal (hunk 1 above)
Comment (prakanth97) [reply tests/jballerina-unit-test/src/test/resources/test-src/types/constant/constant-assignment-negative.bal:28]: Is this addressed?

**Review by prakanth97 — APPROVED** [review]
LGTM

**Review by hasithaa — APPROVED** [review]

**Pull request merged**